src/interface/variable_pack.hpp

```diff
@@ -49,6 +49,53 @@ using SwarmVarList = std::forward_list<std::shared_ptr<ParticleVariable<T>>>;
 // the pairs represent interval (inclusive) of those indices
 using IndexPair = std::pair<int, int>;
 
+// Used for storing the shapes of variable fields
+using Shape = std::vector<int>;
+
+// Index arbitrary rank fields into flattened indices in a VariablePack
+class FlatIdx {
+ public:
+  FlatIdx(std::vector<int> shape, int offset)
+      : shape_("shape", shape.size()), offset_(offset) {
+    for (int i = 0; i < shape.size(); ++i) {
+      shape_(i) = shape[i];
+    }
```

Comment:
This isn't going to work on GPU as is. You'll need to copy the vector into a host mirror and then do a Kokkos deep copy. Alternatively, you could make your `shape` array above the Kokkos view object and do the deep copy outside the `FlatIdx` object.

There's a trade-off here. If the deep copy is performed when the `PackIndexMap` is filled, then it will happen for every variable every time a pack is constructed. (Fortunately packs are cached, so when you ask for a pack it's only reconstructed if something has changed.)

On the other hand, if the deep copy is performed when the `FlatIdx` object is requested, then there will be a small deep copy outside every relevant kernel where the `FlatIdx` object is needed. But it wont' be called when `PackVariables` is called. So that could be advantageous if there aren't many places where `FlatIdx` is needed.

I don't know. What do you think?

Reply:
I was worried about that... 

My guess is that `FlatIdx` won't be too widely needed. It is just useful for radiation moments that have vector indices, species indices and maybe an energy index if we were to ever do multi-group (and maybe a pressure tensor). Therefore, I would say that the deep copy should be done in the FlatIdx constructor. That being said, I have limited experience with both Parthenon and Phoebus so I really don't know how many other use cases there are likely to be.

Reply:
Doing it in the constructor works for me. We can modify it later if it's a performance issue.

Reply:
Since the maximum rank we will have to deal with is 3, I decided to replace the `ParArray1D<int>` with a fixed length `C` array inside `FlatIdx` to keep track of the shape so that instances of `FlatIdx` can be captured by the Kokkos lambdas during parallel dispatch. This seems to work fine and be in line with how indexing is being done in other places in the code and keeps things simple.

Reply:
Yep that should work. :+1: